### ISSUE TITLE
file:write() handles only the first argument and transforms it.

### ISSUE THREAD

**Issue by fperrad** [fengari-lua/fengari-node-cli#6]
`file:write()` handles only the first argument and transforms it.

```
$ fengari
Fengari 0.1.0  Copyright (C) 2017-2018 B. Giannangeli, Daurnimator
Based on: Lua 5.3.4  Copyright (C) 1994-2017 Lua.org, PUC-Rio
> io.stdout:write 'foo'
102,111,111null
> s = 'foo'; print(table.concat({string.byte(s, 1, #s)}, ','))
102,111,111
> io.stdout:write(1, 2)
49null
> io.stdout:write('a', 'b')
97null

$ node -v
v6.11.4
```

**Comment by daurnimator** [comment]
Heh. after #5 where node 6's `fs.readSync` doesn't work with `Uint8Array`s, of course `fs.writeSync` doesn't either. 

https://nodejs.org/docs/latest-v6.x/api/fs.html#fs_fs_writesync_fd_buffer_offset_length_position

Btw, both of these issues have been with the fengari core (https://github.com/fengari-lua/fengari/). This repository is just the node executable: the equivalent of lua.c and luac.c. I wish github had a way to migrate issues within an org.

**Comment by daurnimator** [comment]
Will probably need to fix `package.searchpath` (which uses `fs.openSync`... though `luaL_loadfilex` seemed to work passing a Uint8Array to it?) and the `os` library too (e.g. `os.rename` uses `fs.renameSync`).

**Comment by fperrad** [comment]
with this fix, I got:
```
$ fengari
Fengari 0.1.1  Copyright (C) 2017-2018 B. Giannangeli, Daurnimator
Based on: Lua 5.3.4  Copyright (C) 1994-2017 Lua.org, PUC-Rio
> io.stdout:write 'foo\n'
null

$ lua5.3
Lua 5.3.3  Copyright (C) 1994-2016 Lua.org, PUC-Rio
> io.stdout:write 'foo\n'
foo
file (0x7f2af400b720)
> io.stdin:write 'foo\n'
nil	Bad file descriptor	9
```

writing in `io.stdout` seems to generate an error, but the additionnal message is missing.
access denied on `io.stdout` is expected in a browser, but not with nodejs.

see https://www.lua.org/manual/5.3/manual.html#pdf-file:write

**Comment by daurnimator** [comment]
Are you using the new (unreleased) code?

```
$ pushd $(mktemp -d)
/tmp/tmp.HPgEkLEJXd ~
$ yarn add fengari-lua/fengari fengari-node-cli # uses git HEAD of fengari
yarn add v1.5.1
warning package.json: No license field
warning No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 2 new dependencies.
info Direct dependencies
├─ fengari-node-cli@0.1.0
└─ fengari@0.1.1
info All dependencies
├─ fengari-node-cli@0.1.0
└─ fengari@0.1.1
warning No license field
Done in 4.40s.
$ node -v
v9.11.1
$ fengari
Fengari 0.1.1  Copyright (C) 2017-2018 B. Giannangeli, Daurnimator
Based on: Lua 5.3.4  Copyright (C) 1994-2017 Lua.org, PUC-Rio
> io.stdout:write 'foo\n'
foo
file ([object Object])
> ^C
$ nvm use 6
Now using node v6.14.1 (npm v3.10.10)
$ fengari
Fengari 0.1.1  Copyright (C) 2017-2018 B. Giannangeli, Daurnimator
Based on: Lua 5.3.4  Copyright (C) 1994-2017 Lua.org, PUC-Rio
> io.stdout:write 'foo\n'
foo
file ([object Object])
> ^C
```

**Comment by fperrad** [comment]
Yes, I use your  unreleased code.

`file:write()` is fixed with node **v6.14.1**, but not with node **v6.11.4** (which is shipped with Ubuntu 17.10).
compatibility in nodejs seems to be a nightmare.
support of the latest v6 is enough.

**Comment by daurnimator** [comment]
> file:write() is fixed with node v6.14.1, but not with node v6.11.4 (which is shipped with Ubuntu 17.10).
compatibility in nodejs seems to be a nightmare.

Sigh. You're right.

I installed 6.11.4, the `null` you're seeing is actually an error in error handling. The error handler is failing internally with:
```
TypeError: Cannot read property 'message' of undefined
    at luaL_fileresult (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/lauxlib.js:296:45)
    at g_write (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/liolib.js:152:17)
    at f_write (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/liolib.js:162:12)
    at Object.luaD_precall (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/ldo.js:169:21)
    at Object.luaV_execute (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/lvm.js:530:25)
    at luaD_call (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/ldo.js:367:13)
    at Object.luaD_callnoyield (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/ldo.js:686:5)
    at f_call (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/lapi.js:832:9)
    at luaD_rawrunprotected (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/ldo.js:403:9)
    at Object.luaD_pcall (/tmp/tmp.YWMW3lPJMI/node_modules/fengari/src/ldo.js:665:18)
```

This is a bug in fengari: there is no guarantee that `e` exists.



The error it has originally is due to the written length not being the length of the buffer. That can be replicated easily in plain JS:

```
$ node
> s=new Buffer("foo\n"); console.log(fs.writeSync(1, s))
0
undefined
```

Passing the length seems to make it work:
```
> s=new Buffer("foo\n"); console.log(fs.writeSync(1, s, 0, s.length))
foo
4
undefined
```

This seems to be a node 6.11 quirk; but happy to add in the workaround for now.

**Comment by daurnimator** [comment]
Should be fixed with https://github.com/fengari-lua/fengari/compare/e2bea197494e...53cba4564fba

Thanks for following up!
